src/libraries/Common/src/Interop/Linux/procfs/Interop.ProcFsStat.ParseMapModules.cs

```diff
@@ -0,0 +1,145 @@
+using System.Text;
+
+internal static partial class Interop
+{
+    internal static partial class procfs
+    {
+        private const string MapsFileName = "/maps";
+
+        private static string GetMapsFilePathForProcess(int pid)
+        {
+            return RootPath + pid.ToString(CultureInfo.InvariantCulture) + MapsFileName;
+        }
+
+        private static (long Start, int Size) TryParseAddressRange(string s, ref int start, ref int end)
+        {
+            int pos = s.IndexOf('-', start, end - start);
+            if (pos > 0)
+            {
+                if (long.TryParse(s.AsSpan(start, pos), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out long startingAddress) &&
+                    long.TryParse(s.AsSpan(pos + 1, end - (pos + 1)), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out long endingAddress))
+                {
+                    return (startingAddress, (int)(endingAddress - startingAddress));
+                }
+            }
+
+            return default;
+        }
+
+        private static bool HasReadAndExecFlags(string s, ref int start, ref int end)
+        {
+            bool sawRead = false, sawExec = false;
+            for (int i = start; i < end; i++)
+            {
+                if (s[i] == 'r')
+                    sawRead = true;
+                else if (s[i] == 'x')
+                    sawExec = true;
+            }
+
+            return sawRead & sawExec;
+        }
+
+        internal static ProcessModuleCollection? ParseMapsModules(int pid)
+        {
+            try
+            {
+                return ParseMapsModulesCore(File.ReadLines(GetMapsFilePathForProcess(pid)));
+            }
+            catch (IOException) { }
+            catch (UnauthorizedAccessException) { }
+
+            return null;
+        }
+
+        private static ProcessModuleCollection ParseMapsModulesCore(IEnumerable<string> lines)
+        {
+            Debug.Assert(lines != null);
+
+            ProcessModule? module = null;
+            ProcessModuleCollection modules = new(capacity: 0);
+            bool moduleHasReadAndExecFlags = false;
+
+            foreach (string line in lines)
+            {
+                // Use a StringParser to avoid string.Split costs
+                var parser = new StringParser(line, separator: ' ', skipEmpty: true);
+
+                // Parse the address start and size
+                (long Start, int Size) addressRange = parser.ParseRaw(TryParseAddressRange);
+
+                if (addressRange == default)
+                {
+                    Commit();
+                    continue;
+                }
+
+                // Parse the permissions
+                bool lineHasReadAndExecFlags = parser.ParseRaw(HasReadAndExecFlags);
+
+                // Skip past the offset, dev, and inode fields
+                parser.MoveNext();
+                parser.MoveNext();
+                parser.MoveNext();
+
+                // Parse the pathname
```

Comment:
Nit: it looks like this comment should move down to line 101?

Reply:
Agreed, it was like that before. I think here we can use:

```suggestion
                // we only care about the named modules
```
and move the old comment down.